Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h

```diff
@@ -3038,15 +3038,65 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
     {
-        ASSERT(scalarTypeIsFloatingPoint(simdInfo.lane));
-        ASSERT(simdInfo.signMode != SIMDSignMode::None);
+        ASSERT(supportsAVXForSIMD());
+        ASSERT_UNUSED(simdInfo, simdInfo.signMode == SIMDSignMode::Signed);
         ASSERT(simdInfo.lane == SIMDLane::f32x4);
-        UNUSED_PARAM(input); UNUSED_PARAM(dest); UNUSED_PARAM(simdInfo);
-        // FIXME: Need to support
-        // i32x4.trunc_sat_f32x4_s(a: v128) -> v128
-        // i32x4.trunc_sat_f32x4_u(a: v128) -> v128
+
+        // The instruction cvttps2dq only saturates overflows to 0x80000000 and cannot handle NaN.
+        // However, i32x4.nearest_sat_f32x4_s requires:
+        //     1. saturate positive-overflow integer to 0x7FFFFFFF
+        //     2. saturate negative-overflow integer to 0x80000000
+        //     3. convert NaN or -0 to 0.
+
+        m_assembler.vmovaps_rr(src, scratchFPR1);                               // scratchFPR1 = src
+        m_assembler.vcmpunordps_rrr(scratchFPR1, scratchFPR1, scratchFPR1);     // scratchFPR1 = NaN mask by unordered comparing
+        m_assembler.vandnps_rrr(src, scratchFPR1, scratchFPR1);                 // scratchFPR1 = src with NaN lanes cleared
+        m_assembler.vcvttps2dq_rr(scratchFPR1, scratchFPR1);                    // convert scratchFPR1 to integer with overflow saturated to 0x80000000
+
+        int32_t n = bitwise_cast<int32_t>(0x1.0p+31f);
+        move(TrustedImm32(n), tmp);
+        move32ToFloat(tmp, scratchFPR2);
+        vectorSplatFloat32(scratchFPR2, scratchFPR2);                           // scratchFPR2 = minimum positive-overflow integer 0x80000000
+        m_assembler.vcmpleps_rrr(src, scratchFPR2, scratchFPR2);                // scratchFPR2 = positive-overflow mask by checking scratchFPR2 <= src
+        m_assembler.vpxor_rrr(scratchFPR2, scratchFPR1, dest);                  // convert positive-overflow lane to 0x7FFFFFFF
+    }
+
+    void vectorTruncSatUnsignedFloat32(FPRegisterID src, FPRegisterID dest, RegisterID tmp, FPRegisterID scratchFPR1, FPRegisterID scratchFPR2)
+    {
+        // https://github.com/WebAssembly/simd/pull/247
+        // https://github.com/WebAssembly/relaxed-simd/issues/21
+
+        // The instruction cvttps2dq only saturates overflows to 0x80000000 and cannot handle NaN.
+        // However, i32x4.nearest_sat_f32x4_u requires:
+        //     1. saturate positive-overflow integer to 0xFFFFFFFF
+        //     2. saturate negative-overflow integer to 0
+        //     3. convert NaN or -0 to 0.    
+        
+        m_assembler.vxorps_rrr(scratchFPR1, scratchFPR1, scratchFPR1);
+        m_assembler.vmaxps_rrr(scratchFPR1, src, dest);                     // dest = f[lane]x4 = src with NaN and negatives cleared
+
+        int32_t n = bitwise_cast<int32_t>(2147483647.0f);
+        move(TrustedImm32(n), tmp);
+        move32ToFloat(tmp, scratchFPR1);
+        vectorSplatFloat32(scratchFPR1, scratchFPR1);                       // scratchFPR1 = f[0x80000000]x4
```

Comment:
We should make it load from memory with `static constexpr floats[] = { ... }`. Since emitting these instructions are (1) costly in terms of machine code size, and (2) move32ToFloat is not so cheap.
See the other existing code like `vectorTruncSatSignedFloat64`.

Reply:
You can materialize an address in scratchGPR and use vsubps_mrr for example.